### PR TITLE
Add support for model view editor

### PR DESCRIPTION
**Pull request by AbbiePetcht**

### CODE REVIEW

File: src/sql/workbench/api/node/mainThreadModelViewDialog.ts [hunk 1]
```diff
@@ -32,6 +39,15 @@ export class MainThreadModelViewDialog implements MainThreadModelViewDialogShape
 		throw new Error('Method not implemented.');
 	}
 
+	public $openEditor(modelViewId: string, title: string, position?: vscode.ViewColumn): Thenable<void> {
+		let input = new ModelViewInput(title, modelViewId);
+		let editorOptions = Object.assign({
```
Comment (MattIrv) [inline src/sql/workbench/api/node/mainThreadModelViewDialog.ts]: I think this is left over from my original code but you can remove `Object.assign` here and just leave the inner object since we're not exposing a way for the user to set the editor options (maybe we should in the future?)

File: src/sql/workbench/api/node/mainThreadModelViewDialog.ts [hunk 2]
```diff
@@ -32,6 +39,15 @@ export class MainThreadModelViewDialog implements MainThreadModelViewDialogShape
...
+			preserveFocus: true,
+			pinned: true
+		} as IEditorOptions);
+		return this._editorService.openEditor(input, editorOptions, position as any).then(() => undefined);
```
Comment (MattIrv) [inline src/sql/workbench/api/node/mainThreadModelViewDialog.ts]: Doesn't look like this `then` call is doing anything

File: src/sql/workbench/api/node/extHostModelViewDialog.ts [hunk 3]
```diff
@@ -52,7 +52,26 @@ class ModelViewPanelImpl implements sqlops.window.modelviewdialog.ModelViewPanel
...
+class ViewModelEditorImpl extends ModelViewPanelImpl implements sqlops.workspace.ModelViewEditor {
+	private _content: string;
+	constructor(
+		extHostModelView: ExtHostModelViewShape,
+		private _proxy: MainThreadModelViewDialogShape
+	) {
+		super('modelViewEditor', extHostModelView);
+	}
+
+	public setModelViewId(value: string) {
```
Comment (MattIrv) [inline src/sql/workbench/api/node/extHostModelViewDialog.ts]: I know we do it in the other implementations but I don't think having a private `_content` attribute here that mirrors the private `_modelViewId` attribute from `ModelViewPanelImpl` is the greatest design. I'd suggest just making `_modelViewId` protected on `ModelViewPanelImpl` so that you can access it here

File: samples/sqlservices/src/controllers/mainController.ts [hunk 4]
```diff
@@ -142,6 +146,22 @@ export default class MainController implements vscode.Disposable {
...
+	private openEditor(): void {
+		let editor = sqlops.workspace.createViewModelEditor();
+		editor.registerContent(async view => {
+			let inputBox = view.modelBuilder.inputBox()
+				.withValidation(component => component.value !== 'valid')
+				.component();
+			let formModel = view.modelBuilder.formContainer()
+				.withFormItems([{
+					component: inputBox,
+					title: 'Enter anything but "valid" for build'
```
[the remark this answers is not in the document]
Comment (AbbiePetcht) [reply samples/sqlservices/src/controllers/mainController.ts]: Mean nothing I copied from your code. I will just change it to Enter anthing

---
In reply to: [189377589](https://github.com/Microsoft/sqlopsstudio/pull/1442#discussion_r189377589) [](ancestors = 189377589)

File: samples/sqlservices/src/controllers/mainController.ts [hunk 5]
```diff
@@ -147,19 +147,19 @@ export default class MainController implements vscode.Disposable {
...
+		let editor = sqlops.workspace.createModelViewEditor('Test Editor view');
 		editor.registerContent(async view => {
 			let inputBox = view.modelBuilder.inputBox()
 				.withValidation(component => component.value !== 'valid')
 				.component();
 			let formModel = view.modelBuilder.formContainer()
 				.withFormItems([{
 					component: inputBox,
-					title: 'Enter anything but "valid" for build'
+					title: 'Enter anything'
```
Comment (MattIrv) [inline samples/sqlservices/src/controllers/mainController.ts]: Based on the validation this should still say 'but "valid"'

File: src/sql/workbench/api/node/extHostModelViewDialog.ts (hunk 3 above)
Comment (MattIrv) [reply src/sql/workbench/api/node/extHostModelViewDialog.ts]: Thanks for taking care of this. No need to have the overriden `setModelViewId` method here now since it's not doing anything different

File: src/sql/workbench/api/node/mainThreadModelViewDialog.ts [hunk 6]
```diff
@@ -40,12 +40,19 @@ export class MainThreadModelViewDialog implements MainThreadModelViewDialogShape
...
+			let input = new ModelViewInput(title, modelViewId);
+			let editorOptions = {
+				preserveFocus: true,
+				pinned: true
+			};
+
+			this._editorService.openEditor(input, editorOptions, position as any).then(() => {
+				resolve();
+			}, error => {
+				reject();
```
Comment (MattIrv) [inline src/sql/workbench/api/node/mainThreadModelViewDialog.ts]: This should be `reject(error)` so that the error doesn't get lost.

In any case why the change from before? If your `then` callback doesn't have an error handler then that promise will just be rejected

File: src/sql/workbench/api/node/extHostModelViewDialog.ts (hunk 3 above)
Comment (MattIrv) [reply src/sql/workbench/api/node/extHostModelViewDialog.ts]: Don't see this addressed

File: src/sql/workbench/api/node/mainThreadModelViewDialog.ts [hunk 7]
```diff
@@ -32,6 +39,22 @@ export class MainThreadModelViewDialog implements MainThreadModelViewDialogShape
...
+				preserveFocus: true,
+				pinned: true
+			};
+
+			this._editorService.openEditor(input, editorOptions, position as any).then(() => {
+				resolve();
+			}, error => {
+				reject(error);
+			});
+		});
```
Comment (MattIrv) [inline src/sql/workbench/api/node/mainThreadModelViewDialog.ts]: What I meant with my previous comment is that this code should work but I think it can be simplified to the following without losing anything:

```
let input = new ModelViewInput(title, modelViewId);
let editorOptions = {
    preserveFocus: true,
    pinned: true
};
 
return this._editorService.openEditor(input, editorOptions, position as any).then(() => undefined);
```
